_posts/2021-04-16-rustc-ios-and-an-m1.md

```diff
@@ -0,0 +1,177 @@
+You still can't put code compiled for `arm64-apple-ios` and `arm64-apple-ios14.0-simulator` into the same universal binary[^5],
+because you can have each architecture only once (the `arm64` part in there).
+That's what Carthage and others stumbled over.
+
+Again Apple prepared for that and for a long time they want you to use [XCFramework bundles][XCFramework bundles][^6].
+Carthage just didn't used to support that.
+The [0.37.0 release][carthage release] fixed that.
+
+That still leaves Rust behind, as it doesn't know the new `-simulator` target.
+But as always the Rust community is ahead of the game and [deg4uss3r] started adding a new target in [Rust PR #81966][rust sim target].
+He got half way there when I jumped in to push it over the finish line.
+How these targets work and how LLVM picks the right things to put into the compiled artifacts is severly underdocumented,
+so I had to go the trial-and-error route in combination with looking at LLVM source code to find the missing pieces.
+Turns out: the `14.0` in `arm64-apple-ios14.0-simulator` is actually important.
+
+With the last missing piece in place, the new Rust target landed in February and is available in Nightly.
+Contrary to the main `aarch64-apple-darwin` or `aarch64-apple-ios` target, the simulator target is not Tier 2 yet
+and thus no prebuilt support is available.
+`rustup target add aarch64-apple-darwin` does _not_ work right now.
+I am now in discussions to [promote it to Tier 2][rust tier2 promotion],
+but it's currently blocked by the [RFC: Target Tier Policy].
+
+It works on nightly however and in combination with another cargo capability I'm able to build libraries for the M1 iOS simulator:
+
+```
+cargo +nightly build -Z build-std --target aarch64-apple-ios-sim
+```
+
+For now Glean iOS development on an M1 is possible, but [requires Nightly][glean arm64].
+Goal achieved, I can actually work with this!
+
+In a future blog post I want to explain in more detail how to teach Xcode about all the different targets it should build native code for.
+
+
+## All The Other Projects
+
+This was marked a stretch goal for a reason.
+This involves all the other teams with Rust code and the iOS teams too.
+We're not there yet and there's currently no explicit priority to make development of Firefox iOS on M1 hardware possible.
+But when it comes to it, Glean will be ready for it and the team can assist others to get it over the finish line.
+
+---
+
+Want to hear more about Glean and our cross-platform Rust development?
+Come to next week's [Rust Linz meetup][Rust Linz], where I will be talking about this.
+
+---
+
+_Footnotes:_
+
+[^1]: See [Platform Support] for what the Tiers means.  
+[^2]: The other name for that target.  
+[^3]: "Apple Silicon" is yet another name for what is essentially the same as "M1" or "macOS aarch64"  
+[^4]: Or `arm64` for that matter. Yes, yet another name for the same thing.  
```

Comment:
Might be fun to list all the "functionally identical, but syntactically distinct" ways of saying the same things that you discovered in this journey. 

It'd certainly make a fun slide in a talk.

Reply:
https://asahilinux.org/2021/03/progress-report-january-february-2021/#quick-terminology-note